### PR TITLE
Remove notification on unshare and add type hints

### LINKED ISSUE #2932: Delete notification on unshare
- Share a board to a group by accident
- Unshare it
- See that the notifications to the group users are still around

### PR DESCRIPTION
**Pull request by juliushaertl**
Remove notifications for board shares that are not dismissed yet when unsharing a board again, so there are no stale notifications left and add some typings to the NotificationHelper class.

Fixes #2932

### CODE REVIEW

File: lib/Notification/NotificationHelper.php [hunk 1]
```diff
@@ -157,27 +169,41 @@ public function sendCardAssigned($card, $userId) {
...
 		}
 		if ($acl->getType() === Acl::PERMISSION_TYPE_GROUP) {
 			$group = $this->groupManager->get($acl->getParticipant());
+			if ($group === null) {
+				return;
+			}
 			foreach ($group->getUsers() as $user) {
 				$notification = $this->generateBoardShared($board, $user->getUID());
-				$this->notificationManager->notify($notification);
+				$notification->setDateTime(new DateTime());
```
Comment (nickvergessen) [inline lib/Notification/NotificationHelper.php:196]: Should only be set in the `notify()` branch?

Comment (juliushaertl) [reply lib/Notification/NotificationHelper.php]: Aye, indeed. Thanks :+1:

File: lib/Notification/NotificationHelper.php [hunk 2]
```diff
@@ -157,27 +169,41 @@ public function sendCardAssigned($card, $userId) {
...
 		if ($acl->getType() === Acl::PERMISSION_TYPE_GROUP) {
 			$group = $this->groupManager->get($acl->getParticipant());
+			if ($group === null) {
+				return;
+			}
 			foreach ($group->getUsers() as $user) {
 				$notification = $this->generateBoardShared($board, $user->getUID());
-				$this->notificationManager->notify($notification);
+				if ($markAsRead) {
+					$this->notificationManager->markProcessed($notification);
```
Comment (nickvergessen) [inline lib/Notification/NotificationHelper.php:197]: mention, due date and other notifications are dropped when showing them I guess?

**Review by nickvergessen — APPROVED** [review]

File: lib/Notification/NotificationHelper.php (hunk 2 above)
Comment (juliushaertl) [reply lib/Notification/NotificationHelper.php:210]: Went through them as well now and made sure that they are properly removed on actions that require it.

**Pull request merged**